src/Elastic.Apm.Azure.Functions/ElasticApmMiddleware.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Elastic.Apm.Api;
+using Elastic.Apm.Cloud;
+using Elastic.Apm.Helpers;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Http;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using TraceContext = Elastic.Apm.DistributedTracing.TraceContext;
+
+namespace Elastic.Apm.Azure.Functions;
+
+public class ElasticApmMiddleware : IFunctionsWorkerMiddleware
+{
+	private static readonly IApmLogger Logger;
+	private static readonly string FaasIdPrefix;
+	private static int ColdStart = 1;
+
+	static ElasticApmMiddleware()
+	{
+		Logger = Agent.Instance.Logger.Scoped(nameof(ElasticApmMiddleware));
+		var metaData = new AzureFunctionsMetadataProvider(Logger,
+			new EnvironmentVariables(Logger).GetEnvironmentVariables()).GetMetadata();
+		FaasIdPrefix =
+			$"/subscriptions/{metaData?.Account?.Id}/resourceGroups/{metaData?.Project}/providers/Microsoft.Web/sites/{metaData?.Instance}/functions/";
+		Logger.Trace()?.Log($"FaasIdPrefix: {FaasIdPrefix}");
+	}
+
+	private static bool IsColdStart() => Interlocked.Exchange(ref ColdStart, 0) == 1;
+
+	public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
+	{
+		try
+		{
+			Logger.Trace()?.Log($"{nameof(Invoke)} - {context.FunctionDefinition.Name}");
+
+			var data = GetTriggerSpecificData(context);
+			await Agent.Tracer.CaptureTransaction(data.Name, ApiConstants.TypeRequest, async t =>
+			{
+				if (t is Transaction transaction) // TODO: Expose FaaS property on ITransaction?
+				{
+					transaction.FaaS = new Faas
+					{
+						Name = context.FunctionDefinition.Name,
+						Id = $"{FaasIdPrefix}{context.FunctionDefinition.Name}",
+						Trigger = new Trigger { Type = data.TriggerType },
+						Execution = context.InvocationId,
+						ColdStart = IsColdStart()
+					};
+				}
+
+				await next(context);
+
+				t.Outcome = Outcome.Success;
+				t.Result = GetTriggerSpecificResult(context);
```

Comment:
We have something similar [here](https://github.com/elastic/apm-agent-dotnet/blob/main/src/Elastic.Apm/Model/Transaction.cs#LL793) - we could maybe just call that.

Reply:
The idea was to also support other trigger types (non-HTTP) here over time, but thanks for the hint. I will reuse the existing code for HTTP.

Reply:
> The idea was to have this dedicated method so we can support other trigger types (non-HTTP) in there over time. 

That's a good point, and honestly during my review I was very much focused on HTTP. The `SetTriggerSpecificResult` you added handled this nicely.